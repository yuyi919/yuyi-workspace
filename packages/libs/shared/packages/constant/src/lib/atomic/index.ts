export * from "./keys";
export * from "./function";
export * from "./constructor";
