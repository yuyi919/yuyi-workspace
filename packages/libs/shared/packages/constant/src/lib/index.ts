import { ArrayIterator, Types } from "@yuyi919/shared-types";
import {
  ARRAY,
  CALLER,
  CREATE_NEW,
  FUNCTION,
  KEY_PROTOTYPE,
  OBJECT,
  PROMISE,
  REGEXP
} from "./atomic";

export function CREATE_PROMISE<T>(
  executor: (resolve: (value?: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void
): Promise<T> {
  return CREATE_NEW(PROMISE, [executor]);
}
export function EMPTY_PROMISE<T = any>(value: T) {
  return PROMISE.resolve(value);
}

export function BIND<
  T extends Types.Function.Base<[any, ...any[]], any>,
  TA,
  R extends T extends (this: any, ...other: infer K) => infer Re
    ? Types.Function.Base<K, Re>
    : Function
>(Func: T, thisArgs: TA, ...args: any[]): R {
  // @ts-ignore
  return FUNCTION.bind.call(Func, thisArgs, ...args);
}

// tslint:disable-next-line: class-name
// export type STATIC_BIND<
//   T extends Function,
//   Bindex extends any[]
// > = {

// }
export function BindArg$$<T extends Function, A extends Types.Function.Arg0<T>>(
  Func: T,
  arg0: A
): Types.Function.Shift1<T, A>;
export function BindArg$$<
  T extends Function,
  A extends Types.Function.Arg0<T>,
  B extends Types.Function.Arg1<T>
>(Func: T, arg0: A, arg1: B): Types.Function.Shift2<T, A, B>;
export function BindArg$$<
  T extends Function,
  A extends Types.Function.Arg0<T>,
  B extends Types.Function.Arg1<T>,
  C extends Types.Function.Arg2<T>
>(Func: T, arg0: A, arg1: B, arg2: C): Types.Function.Shift3<T, A, B, C>;
export function BindArg$$(Func: any, ...args: any[]): any {
  // @ts-ignore
  return BIND(Func, null, ...args);
}

export function CREATE_OBJECT_IS(value: any): (value: any) => boolean {
  return BindArg$$(OBJECT.is, value);
}

export interface FunctionArgsShifter<
  T extends Function,
  P extends T["prototype"],
  K extends keyof P
> {
  <Args extends P[K] extends (arg: infer A) => any ? A : any>(source: P, arg: Args): P[K] extends (
    ...args: any[]
  ) => infer R
    ? R
    : any;
}

export function INSTANCE_BIND<T extends Function, P extends T["prototype"], K extends keyof P>(
  instance: T,
  key: K
): FunctionArgsShifter<T, P, K> {
  // try {
  return BIND(CALLER, instance[KEY_PROTOTYPE][key]) as any;
  // } catch (error) {
  //   console.log(instance, key);
  // }
}
export function PROTOTYPE_BIND<T extends Function, P extends T["prototype"], K extends keyof P>(
  instance: T,
  key: K
): P[K] {
  // @ts-ignore
  return BIND(instance[KEY_PROTOTYPE][key], instance[KEY_PROTOTYPE]);
}

/**
 * 原生的map循环改为函数调用
 * @param arr
 * @param callbackfn
 */
export const APPLY = (fn: any, args: any[], thisArgs?: any) =>
  fn.apply(thisArgs, args) as {
    <TArgs extends any[], TResult = any, T = any>(
      source: Function,
      args: TArgs,
      thisArgs?: T
    ): TResult;
  };
/**
 * 原生的map循环改为函数调用
 * @param arr
 * @param callbackfn
 */
export const MAP = INSTANCE_BIND(ARRAY, "map") as {
  <T extends any, R = T>(arr: T[], callbackfn: ArrayIterator<T, R>, initialValue?: any[]): R[];
};

export function DEFINED_ESMODULE(_exports: any, desc: any) {
  return Object.defineProperty(_exports, "__esModule", desc);
}
// MAP()
/**
 * 原生的Array.prototype.some改为函数调用
 * @param arr
 * @param callbackfn
 * @param thisArg
 */
export const SOME = INSTANCE_BIND(ARRAY, "some") as {
  <T extends any>(arr: T[], callbackfn: ArrayIterator<T, boolean>, thisArg?: any): boolean;
};
/**
 * 原生的for循环改为函数调用
 * @param arr
 * @param callbackfn
 */
export const FOR_EACH = INSTANCE_BIND(ARRAY, "forEach") as {
  <T extends any>(arr: T[], callbackfn: ArrayIterator<T, any>): void;
};
/**
 * 原生数组的concat改为函数调用
 * @param arr
 * @param callbackfn
 */
export const ARR_CONCAT = PROTOTYPE_BIND(ARRAY, "concat") as {
  <T extends any, A extends any = T>(arr: T[], ...next: A[][]): (A | T)[];
  <T extends any, A extends any = T>(arr: T[], ...next: A[]): (A | T)[];
};
/**
 * 原生数组的silce改为函数调用
 * @param arr
 * @param callbackfn
 */
export const ARR_SLICE = INSTANCE_BIND(ARRAY, "slice") as {
  <T extends any>(arr: T[], startIndex?: number, endIndex?: number): T[];
  // tslint:disable-next-line: unified-signatures
  <T extends any>(arr: IArguments, startIndex?: number, endIndex?: number): T[];
};

/**
 * 原生的for循环改为函数调用
 * @param arr
 * @param callbackfn
 */
export const FILTER = INSTANCE_BIND(ARRAY, "filter") as {
  <T extends any>(arr: T[], callbackfn: ArrayIterator<T, boolean>): T[];
};

/**
 * 原生的reduce改为函数调用
 * @param arr
 * @param callbackfn
 * @param initialValue
 * @example
 * REDUCE([1, 2, 3, 4], (target: any, i: number) => ({ ...target, [i]: i }), {})
 */
export const REDUCE = INSTANCE_BIND(ARRAY, "reduce") as {
  <T extends any, R = T>(
    arr: T[],
    callbackfn: (previousValue: R, currentValue: T, currentIndex: number, array: T[]) => R,
    initialValue?: R
  ): R;
};
/**
 * 原生的Array<T>.push改为函数调用
 * @param arr 目标数组
 * @param elements 新元素
 * @example
 * ARR_PUSH([1, 2, 3, 4], 5, 6, 7)
 */
export const ARR_PUSH = INSTANCE_BIND(ARRAY, "push") as {
  <T extends any, E = T>(arr: T[], ...elements: E[]): number;
};

/**
 * 原生的push改为函数调用
 * @param arr 目标数组
 * @param elements 新元素
 * @example
 * REGEXP_TEST(/123/, '123') // => true
 */
export const REGEXP_TEST = INSTANCE_BIND(REGEXP, "test") as {
  (reg: RegExp, str: string): boolean;
};

export const delay$$: <P extends any[]>(
  handle: (...args: P) => void,
  sec: number,
  ...args: P
) => number = BIND(setTimeout, null) as any;

/**
 *
 * @param arr
 * @param callbackfn
 */
export function MAP$$<T, R>(arr: T[], callbackfn: ArrayIterator<T, R>): R[];
export function MAP$$<T, R>(
  arr: T[],
  callbackfn: ArrayIterator<T, R>,
  initialValue: R[] = [],
  i = 0,
  length = arr.length - 1
) {
  initialValue[i] = callbackfn(arr[i], i, arr);
  return i < length
    ? // @ts-ignore
      MAP$$(arr, callbackfn, initialValue, i + 1, length)
    : initialValue;
}

function FOR_EACH_P$$<T>(
  arr: T[],
  callbackfn: ArrayIterator<T, any>,
  tmp: { i: number },
  length: number
) {
  callbackfn(arr[tmp.i], tmp.i, arr);
  // console.trace('test')
  // @ts-ignore
  return tmp.i++ < length ? FOR_EACH_P$$(arr, callbackfn, tmp, length) : true;
}
/**
 *
 * @param arr
 * @param callbackfn
 */
export function FOR_EACH$$<T>(arr: T[], callbackfn: ArrayIterator<T, any>): true {
  const tmp = { i: 0 };
  const length = arr.length - 1;
  try {
    return FOR_EACH_P$$(arr, callbackfn, tmp, length);
  } catch (error) {
    while (length - tmp.i++) callbackfn(arr[tmp.i], tmp.i, arr);
    return true;
  }
}

//
export const Key_useDeprecatedSynchronousErrorHandling$$ = "useDeprecatedSynchronousErrorHandling";

export * from "./atomic";
