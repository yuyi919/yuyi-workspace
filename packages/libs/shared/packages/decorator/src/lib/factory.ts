/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Types } from "@yuyi919/shared-types";
import "reflect-metadata";

export type MetaOption =
  | ConstructorMetaOption<any[], any>
  | PropertyMetaOption<any[], any>
  | MethodMetaOption<any[], any>;
export type MetaKinds = MetaOption["kind"];
export interface ConfigureMetaOption<S extends any[], T> {
  config?: (
    meta: {
      target: Types.ConstructorType<any>;
      propertyKey?: string;
      descriptor?: TypedPropertyDescriptor<any>;
    },
    ...param: S
  ) => T;
}
export interface ConstructorMetaOption<S extends any[], T> extends ConfigureMetaOption<S, T> {
  kind: "constructor";
  children?: string;
}
export interface PropertyMetaOption<S extends any[], T> extends ConfigureMetaOption<S, T> {
  kind: "property";
}
export interface MethodMetaOption<S extends any[], T> extends ConfigureMetaOption<S, T> {
  kind: "method";
}
export type MetaConfig<T, Name extends string = string, Kind extends MetaKinds = MetaKinds> = {
  name: Name;
  meta: T;
  kind: Kind;
};
function _appendMeta<T>(
  MetaKey: any,
  target: Types.Recordable,
  kind: MetaKinds,
  meta: T,
  propertyKey: string
) {
  const arr: MetaConfig<T>[] = _getMeta(MetaKey, target) || [];
  arr.push({ name: propertyKey, meta, kind });
  Reflect.defineMetadata(MetaKey, arr, target);
}
function _defineMeta<T>(
  MetaKey: any,
  target: Types.ConstructorType<any>,
  meta: T,
  kind: MetaKinds,
  propertyKey?: string
): void {
  Reflect.defineMetadata(
    MetaKey,
    { name: propertyKey || target.name, meta, kind } as MetaConfig<T>,
    target,
    propertyKey
  );
}

function _getMeta(MetaKey: any, target: Types.Recordable, propertyKey?: string): any {
  if (propertyKey) {
    return (Reflect.getMetadata(MetaKey, target) as MetaConfig<any>[])?.find(
      (o) => o.name === propertyKey
    );
  }
  return Reflect.getMetadata(MetaKey, target);
}

/**
 * 创建一组静态元数据的装饰器
 * @param configure 配置项集合
 * @returns
 */
export function createStaticMetaDataDecorators<
  Namespace extends string,
  Config extends Record<string, MetaOption>
>(namespace: Namespace, configure: Config) {
  const metaKeys = {} as Record<keyof Config, symbol>;
  const decorators = {} as {
    [K in keyof Config]: Config[K] extends ConstructorMetaOption<infer S, any>
      ? (...options: S) => ClassDecorator
      : Config[K] extends PropertyMetaOption<infer S, any>
      ? (...options: S) => PropertyDecorator
      : PropertyDecorator;
  };
  function getMeta<K extends keyof Config, Target extends Types.ConstructorType<any>>(
    key: K,
    target: Target
  ): Config[K] extends ConfigureMetaOption<any[], infer T>
    ? Config[K] extends ConstructorMetaOption<any, any> // 根据装饰器类型
      ? MetaConfig<T, string, Config[K]["kind"]>
      : MetaConfig<T, Types.KeyOf<InstanceType<Target>>, Config[K]["kind"]>[]
    : void;
  function getMeta<
    K extends keyof Config,
    Target extends Types.ConstructorType<any>,
    PropertyKey extends keyof InstanceType<Target>
  >(
    key: K,
    target: Target,
    propertyKey: PropertyKey
  ): Config[K] extends ConfigureMetaOption<any[], infer T>
    ? MetaConfig<T, PropertyKey & string, Config[K]["kind"]>
    : void;
  function getMeta(key: keyof Config, target: any, propertyKey?: string): any {
    return _getMeta(metaKeys[key], target, propertyKey);
  }
  const helpers = {
    getMeta,
  };
  for (const key in configure) {
    const meta = configure[key];
    const { kind } = meta;
    const metaKey = (metaKeys[key] = Symbol(namespace + "_meta:" + key));
    if (kind === "property") {
      decorators[key] = function (...args: any[]) {
        return function (target, propertyKey: string) {
          const propertyOpts = meta.config(
            { target: target.constructor as Types.ConstructorType<any>, propertyKey },
            ...args
          );
          _appendMeta(metaKey, target.constructor, kind, propertyOpts, propertyKey);
        } as PropertyDecorator;
      } as any;
    } else if (kind === "constructor") {
      decorators[key] = function (...args: any[]) {
        return function (target: Types.ConstructorType<any>) {
          const propertyOpts = meta.config({ target }, ...args);
          _defineMeta(metaKey, target, propertyOpts, kind);
        } as ClassDecorator;
      } as any;
    }
  }
  return {
    ...decorators,
    ...helpers,
    metaKeys,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    transform<T extends any>(plain: any, Target: Types.ConstructorType<T>) {
      const result = { ...plain } as T;
      // for (const $Target of [Target].concat(Targets)) {
      // const keys = Object.getOwnPropertyNames($Target.prototype);
      // console.log($Target.name, keys);
      // merge(
      // result,
      // pickBy(plain, (_, key) => {
      // console.log(plain, key);
      // return keys.includes(key) || keys.includes(key.replace(/^.+?\./, ""));
      // })
      // );
      // }
      return result;
    },
  };
}
