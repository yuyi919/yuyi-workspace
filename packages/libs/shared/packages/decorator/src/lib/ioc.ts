import { postConstruct, injectable as Injectable, tagged, multiInject } from "inversify";
import "reflect-metadata";

export const T = {
  Config: Symbol("Config"),
  Plugin: Symbol("Plugin"),
  NamedPlugin: Symbol("NamedPlugin"),
};

@Injectable()
export class Config {
  constructor(@multiInject(T.Plugin) public plugins: any[]) {}

  @postConstruct()
  loaded() {
    console.log(this.plugins.map((o) => (o instanceof Function ? o() : o)));
  }
}
