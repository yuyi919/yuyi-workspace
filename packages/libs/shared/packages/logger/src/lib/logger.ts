export function sharedPackagesLogger(): string {
  return "shared-packages-logger";
}
