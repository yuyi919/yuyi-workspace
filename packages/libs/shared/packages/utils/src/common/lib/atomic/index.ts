export * from "./custom-is";
export * from "./getRealLength";
export * from "./isEqual";
export * from "./isFunction";
export * from "./isNil";
export * from "./isNumber";
export * from "./isObject";
export * from "./lodash";
export * from "./propertiesDefine";
export * from "./toGetter";
export * from "./helper";

import {
  setTrue$$,
  setValue$$,
  setWith$$,
  setWithEntries$$,
  setWithEntriesReverse$$,
  setWithKeyValue$$,
  setWithLabelValue$$,
  setWithValueLabel$$
} from "./setter";

/**
 * @internal
 */
export const Setter = {
  setTrue$$,
  setValue$$,
  setWith$$,
  setWithEntries$$,
  setWithEntriesReverse$$,
  setWithKeyValue$$,
  setWithLabelValue$$,
  setWithValueLabel$$
};

export {
  is,
  isStr,
  isNil,
  isRegExp,
  isArguments,
  isArr,
  isBigInt,
  isBool,
  isDate,
  isEmpty,
  isError,
  isEsModule,
  isEsModuleWithDefaultExport,
  isFn,
  isNull,
  isNum,
  isNaN,
  isObj,
  isPrimitive,
  isPromise,
  isSymbol,
  isThenable,
  isUndefined,
  hasOwnKey
} from "@yuyi919/shared-types";
