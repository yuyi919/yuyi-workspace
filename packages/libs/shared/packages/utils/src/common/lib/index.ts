/**
 * 实用工具
 * @packageDocumentation
 */

export * from "./atomic";
export * from "./struct";
export * from "./async";

export function extendMap<T extends Map<any, any>>(...map: T[]): T {
  let r = [] as [string, any][];
  for (const o of map) {
    if (o) {
      r = r.concat(Array.from(o.entries()));
    }
  }
  return new Map(r) as T;
}
