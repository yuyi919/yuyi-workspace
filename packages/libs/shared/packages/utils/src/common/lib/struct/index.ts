export * from "./autobind";
export * from "./castUtils";
export * from "./convertArr2Map";
export * from "./createGroupWith";
export * from "./escapeConvert";
export * from "./filterMap";
export * from "./pipe";
export * from "./reduceMap";
export * from "./stub";
export * from "./trackMerge";
export * from "./autobind";
export * from "./getListDifferent";
export * from "./likeArray";
export * from "./zipEmptyData";
export * from "./isSubArray";
export * from "./randomInt";
