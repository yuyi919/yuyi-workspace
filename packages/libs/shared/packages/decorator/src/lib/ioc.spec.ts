import { T, Config } from "@yuyi919/shared-decorator/src/lib/ioc";
import {
  inject,
  decorate,
  injectable as Injectable,
  Container,
  tagged,
  multiBindToService,
  postConstruct,
} from "inversify";

export class PluginA {
  prop?: number;
  constructor() {
    console.log("A");
  }
}
decorate(Injectable(), PluginA);

export class PluginB {
  constructor(public a: any) {
    console.log("B");
  }
}
decorate(tagged("tag", "A"), PluginB, 0);
decorate(inject(T.Plugin), PluginB, 0);
decorate(Injectable(), PluginB);

describe("basic", () => {
  it("aaa", () => {
    const container = new Container({ defaultScope: "Request" });
    container.bind(T.Config).to(Config);
    container.bind(T.Plugin).to(PluginA).whenTargetTagged("tag", "A");
    container.bind(T.Plugin).to(PluginB).whenTargetTagged("tag", "B");
    container.bind(T.Plugin).to(PluginA).whenTargetIsDefault();
    container.bind(T.Plugin).to(PluginB).whenTargetIsDefault();
    expect(container.get(T.Config)).toMatchSnapshot();
  });
});
