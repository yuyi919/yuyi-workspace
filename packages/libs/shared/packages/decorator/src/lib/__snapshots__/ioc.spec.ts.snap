// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic aaa 1`] = `
Config {
  "plugins": Array [
    PluginA {},
    PluginB {
      "a": PluginA {},
    },
  ],
}
`;
